compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1098,66 +1098,135 @@ private BType inferTableMemberType(List<BType> memTypes, BType expType) {
+                if (fieldNameToFields.containsKey(key)) {
+                    fieldNameToFields.get(key).add(field);
+                } else {
+                    fieldNameToFields.put(key, new ArrayList<>() {{
+                        add(field);
+                    }});
+                }
+            }
 
-            if (fieldNames.contains(fieldName)) {
-                dlog.error(tableConstructorExpr.pos,
-                        DiagnosticErrorCode.CANNOT_INFER_MEMBER_TYPE_FOR_TABLE_DUE_AMBIGUITY,
-                        fieldName);
-                return symTable.semanticError;
+            if (!member.sealed) {
+                BType restFieldType = member.restFieldType;
+                if (isUniqueType(restFieldTypes, restFieldType)) {
+                    restFieldTypes.add(restFieldType);
+                }
             }
-            fieldNames.add(fieldName);
+        }
 
-            boolean isOptional = true;
-            for (BField commonField : commonFieldSet) {
-                if (commonField.name.value.equals(fieldName)) {
-                    isOptional = false;
-                    requiredFieldNames.add(commonField.name.value);
+        LinkedHashSet<BField> inferredFields = new LinkedHashSet<>();
+        int memTypesSize = memTypes.size();
+
+        fieldNameToFields.forEach((fieldName, fields) -> {
+            List<BType> uniqueTypes = getUniqueTypesFromSameNameFieldList(fields);
+
+            for (BType memType : memTypes) {
+                BRecordType bMemType = (BRecordType) memType;
+                if (bMemType.sealed) {
+                    continue;
+                }
+
+                if (bMemType.fields.containsKey(fieldName)) {
+                    continue;
+                }
+
+                BType restFieldType = bMemType.restFieldType;
+                if (isUniqueType(uniqueTypes, restFieldType)) {
+                    uniqueTypes.add(restFieldType);
                 }
             }
 
+            BField resultantField = replaceFieldType(fields.get(0), uniqueTypes);
+            boolean isOptional = hasOptionalFields(fields) || fields.size() != memTypesSize;
+
             if (isOptional) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.OPTIONAL));
-            } else if (requiredFieldNames.contains(fieldName) && keySpecifierFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) + Flags.asMask(EnumSet.of(Flag.READONLY));
-            } else if (requiredFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED));
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.OPTIONAL));
+            } else if (keySpecifierFieldNames.contains(fieldName)) {
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) +
+                        Flags.asMask(EnumSet.of(Flag.READONLY));
```

Comment:
```suggestion
                    BVarSymbol symbol = resultantField.symbol;
                    symbol.flags |= Flags.REQUIRED;
                    symbol.flags |= Flags.READONLY;
```

Reply:
Is there any reason we consider previous flags into account here? 
If not isn't it better to not depend on them/assume them.

Reply:
My bad, misread it. Can't we just do

```suggestion
                resultantField.symbol.flags = Flags.REQUIRED | Flags.READONLY;
```